crates/libs/windows/src/core/hstring.rs

```diff
@@ -420,27 +415,27 @@ impl Header {
         // TODO: allow this failure to propagate
         let header = heap_alloc(alloc_size).expect("Could not successfully allocate for HSTRING") as *mut Header;
 
+        // SAFETY: uses `std::ptr::write` (since `header` is unintialized). `Header` is safe to be all zeros.
         unsafe {
-            (*header).flags = 0;
+            header.write(std::mem::MaybeUninit::<Header>::zeroed().assume_init());
             (*header).len = len;
-            (*header).data = &mut (*(*header).shared.as_mut_ptr()).buffer_start;
-            (*(*header).shared.as_mut_ptr()).count = RefCount::new(1);
+            (*header).count = RefCount::new(1);
+            (*header).data = &mut (*header).buffer_start;
```

Comment:
I wonder if `header` should be `Pin<*mut Header>` since it's self-referential? It's probably more trouble than it's worth, since `Pin` is not very ergonomic right now.

Reply:
I agree that `Pin` isn't worth it here. We have to worry about safety with the raw pointer anyway, ensuring that `Header` isn't moved is something we should be able to manually verify in the very few places `*mut Header`s are dereferenced.